mapdataprefetcher/app/src/main/java/com/here/android/example/map/data/prefetcher/MainActivity.java

```diff
@@ -0,0 +1,131 @@
+
+        if (hasPermissions(this, RUNTIME_PERMISSIONS)) {
+            setupMapFragmentView();
+        } else {
+            ActivityCompat
+                    .requestPermissions(this, RUNTIME_PERMISSIONS, REQUEST_CODE_ASK_PERMISSIONS);
+        }
+    }
+
+    /**
+     * Only when the app's target SDK is 23 or higher, it requests each dangerous permissions it
+     * needs when the app is running.
+     */
+    private static boolean hasPermissions(Context context, String... permissions) {
+        if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && permissions != null) {
+            for (String permission : permissions) {
+                if (ActivityCompat.checkSelfPermission(context, permission)
+                        != PackageManager.PERMISSION_GRANTED) {
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
+    @Override
+    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
+                                           @NonNull int[] grantResults) {
+        switch (requestCode) {
+            case REQUEST_CODE_ASK_PERMISSIONS: {
+                for (int index = 0; index < permissions.length; index++) {
+                    if (grantResults[index] != PackageManager.PERMISSION_GRANTED) {
+
+                        /*
+                         * If the user turned down the permission request in the past and chose the
+                         * Don't ask again option in the permission request system dialog.
+                         */
+                        if (!ActivityCompat
+                                .shouldShowRequestPermissionRationale(this, permissions[index])) {
+                            Toast.makeText(this, "Required permission " + permissions[index]
+                                            + " not granted. "
+                                            + "Please go to settings and turn on for sample app",
+                                    Toast.LENGTH_LONG).show();
+                        } else {
+                            Toast.makeText(this, "Required permission " + permissions[index]
+                                    + " not granted", Toast.LENGTH_LONG).show();
+                        }
+                    }
+                }
+
+                setupMapFragmentView();
+                break;
+            }
+            default:
+                super.onRequestPermissionsResult(requestCode, permissions, grantResults);
+        }
+    }
+
+    private void setupMapFragmentView() {
+        // All permission requests are being handled. Create map fragment view. Please note
+        // the HERE Mobile SDK requires all permissions defined above to operate properly.
+        m_mapFragmentView = new MapFragmentView(this);
+    }
+
+    @Override
+    public boolean onCreateOptionsMenu(Menu menu) {
+        getMenuInflater().inflate(R.menu.map_menu, menu);
```

Comment:
Also move it in `MapFragmentView`:

`m_mapFragmentView.onCreateOptionsMenu(menu);`